src/filesystem.rs

```diff
@@ -47,175 +47,305 @@ pub struct FilesystemReader<R: SquashFsReader> {
+                    reader.read_exact(data)?;
+                    Ok(RawDataBlock::Fragment {
+                        data,
+                        fragment: *fragment,
+                    })
+                }
+            },
+        }
+    }
+    pub fn next_block<'b>(
+        &mut self,
+        buf: &'b mut Vec<u8>,
+    ) -> Option<Result<RawDataBlock<'b>, SquashfsError>> {
+        self.current_block
+            .next()
+            .map(|next| self.read_raw_data(buf, next))
+    }
+    fn fragment_range(&self) -> std::ops::Range<usize> {
+        let block_len = self.file.system.block_size as usize;
+        let block_num = self.file.basic.block_sizes.len();
+        let file_size = self.file.basic.file_size as usize;
+        let frag_len = file_size - (block_num * block_len);
+        let frag_start = self.file.basic.block_offset as usize;
+        let frag_end = frag_start + frag_len;
+        frag_start..frag_end
+    }
+    pub fn decompress<'b>(
+        &self,
+        block: &'_ RawDataBlock<'b>,
+        buf: &'b mut Vec<u8>,
+    ) -> Result<&'b [u8], SquashfsError> {
+        match block {
+            RawDataBlock::Block { data, block } if block.uncompressed() => {
+                Ok(data)
+            }
+            RawDataBlock::Block { data, .. } /*if !_block.uncompressed()*/ => {
+                buf.clear();
+                compressor::decompress(data, buf, self.file.system.compressor)?;
+                Ok(buf.as_slice())
+            },
+            RawDataBlock::Fragment { data, fragment } => {
+                let data = if fragment.size.uncompressed() {
+                    data
+                } else {
+                    buf.clear();
+                    compressor::decompress(data, buf, self.file.system.compressor)?;
+                    //store the cache, so decompression is not duplicated
+                    self.file.system.cache
+                        .borrow_mut()
+                        .fragment_cache
+                        .insert(fragment.start, buf.clone());
+                    buf.as_slice()
+                };
+                let range = self.fragment_range();
+                Ok(&data[range])
             },
         }
-        Ok(())
     }
+    pub fn into_reader(self) -> SquashfsReadFile<'a, R> {
+        let bytes_available = self.file.basic.file_size as usize;
+        let buf_read = Vec::with_capacity(self.file.system.block_size as usize);
+        let buf_decompress = Vec::with_capacity(self.file.system.block_size as usize);
+        SquashfsReadFile {
+            raw_data: self,
+            buf_read,
+            buf_decompress,
+            last_read: &[],
+            bytes_available,
+        }
+    }
+}
 
+pub struct SquashfsReadFile<'a, R: ReadSeek> {
+    raw_data: SquashfsRawData<'a, R>,
+    buf_read: Vec<u8>,
+    buf_decompress: Vec<u8>,
+    //Safety: will point to one of the buffers, modify that require extra care
+    last_read: &'static [u8],
```

Comment:
Not at my computer, but couldn't this be replaced with a `Box` for no performance loss and no unsafe?

Or `RefCell`, as I _think_ that would keep mem usage down?

Reply:
Yes, the unsafe can be removed with little to no performance loss. No `Box` or `RefCell` required, just need to replace the reference by an index.

Reply:
The main change being `std::mem::swap(&mut self.buf_read, &mut self.buf_decompress)` so the final data is always stored in `buf_decompress`.

Reply:
Done on [0a36dc1](https://github.com/wcampbell0x2a/backhand/pull/132/commits/0a36dc15e6bdecc7ebf8bc1f2d3719d339926423), with some extra changes to make it possible.